butane_cli/src/lib.rs

```diff
@@ -0,0 +1,265 @@
+        println!("Created migration {name}");
+    } else {
+        println!("No changes to migrate");
+    }
+    Ok(())
+}
+
+pub fn migrate() -> Result<()> {
+    let spec = load_connspec()?;
+    let mut conn = db::connect(&spec)?;
+    let to_apply = get_migrations()?.unapplied_migrations(&conn)?;
+    println!("{} migrations to apply", to_apply.len());
+    for m in to_apply {
+        println!("Applying migration {}", m.name());
+        m.apply(&mut conn)?;
+    }
+    Ok(())
+}
+
+pub fn rollback_to(mut conn: Connection, to: &str) -> Result<()> {
+    let ms = get_migrations()?;
+    let to_migration = match ms.get_migration(to) {
+        Some(m) => m,
+        None => {
+            eprintln!("No such migration!");
+            std::process::exit(1);
+        }
+    };
+
+    let to_unapply = ms.migrations_since(&to_migration)?;
+    if to_unapply.is_empty() {
+        eprintln!("That is the latest migration, not rolling back to anything. If you expected something to happen, try specifying the migration to rollback to.");
+    }
+    for m in to_unapply.into_iter().rev() {
+        println!("Rolling back migration  {}", m.name());
+        m.downgrade(&mut conn)?;
+    }
+    Ok(())
+}
+
+pub fn rollback_latest(mut conn: Connection) -> Result<()> {
+    match get_migrations()?.latest() {
+        Some(m) => {
+            println!("Rolling back migration  {}", m.name());
+            m.downgrade(&mut conn)?;
+        }
+        None => {
+            eprintln!("No migrations applied!");
+            std::process::exit(1)
+        }
+    };
+    Ok(())
+}
+
+pub fn embed() -> Result<()> {
+    let srcdir = std::env::current_dir()?.join("src");
+    if !srcdir.exists() {
+        eprintln!("src directory not found");
+        std::process::exit(1);
+    }
+    let path = srcdir.join("butane_migrations.rs");
+
+    let mut mem_ms = MemMigrations::new();
+    for m in get_migrations()?.all_migrations()? {
+        let mut new_m = mem_ms.new_migration(&m.name());
+        copy_migration(&m, &mut new_m)?;
+        mem_ms.add_migration(new_m)?;
+    }
+    let json = serde_json::to_string(&mem_ms)?;
+
+    let src = format!(
+        "
+use butane::migrations::MemMigrations;
+use std::result::Result;
+pub fn get_migrations() -> Result<MemMigrations, butane::Error> {{
+    let json = r#\"{json}\"#;
+    MemMigrations::from_json(json)
+}}"
+    );
+
+    let mut f = std::fs::File::create(path)?;
+    f.write_all(src.as_bytes())?;
+
+    let mut cli_state = CliState::load()?;
+    cli_state.embedded = true;
+    cli_state.save()?;
+    Ok(())
+}
+
+pub fn load_connspec() -> Result<db::ConnectionSpec> {
+    match db::ConnectionSpec::load(base_dir()?) {
+        Ok(spec) => Ok(spec),
+        Err(butane::Error::IO(_)) => {
+            eprintln!("No Butane connection info found. Did you run butane init?");
+            std::process::exit(1);
+        }
+        Err(e) => Err(e.into()),
+    }
+}
+
+pub fn list_migrations() -> Result<()> {
```

Comment:
This just prints to the console rather than returning anything, so I'm not sure how it's useful for programmatic usage? Perhaps I'm misunderstanding how you're envisioning using this.